### PR TITLE
Seesaw Loss for Long-Tailed Instance Segmentation (CVPR 2021)

### PR DESCRIPTION
**Pull request by myownskyW7**

### CODE REVIEW

File: mmdet/models/losses/seesaw_loss.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    """
+    assert cls_score.size(-1) == num_classes
+    assert len(cum_samples) == num_classes
+
+    onehot_labels = F.one_hot(labels, num_classes)
+    seesaw_weights = cls_score.new_ones(onehot_labels.size())
+
+    # mitigation factor
+    if p > 0:
+        sample_ratio_matrix = cum_samples[None, :].clamp_min(
```
Comment (ZwwWayne) [inline mmdet/models/losses/seesaw_loss.py:49]: `clamp_min ` seems to be a function not listed in the pytorch API documentation. It is a little bit risky to use it. Can we simply use `clamp(min=1)` here?

File: mmdet/models/losses/seesaw_loss.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            1) / cum_samples[:, None].clamp_min(1)
+        index = (sample_ratio_matrix < 1.0).float()
+        sample_weights = sample_ratio_matrix.pow(p) * index + (1 - index)
+        mitigation_factor = sample_weights[labels.long(), :]
+        seesaw_weights = seesaw_weights * mitigation_factor
+
+    # compensation factor
+    if q > 0:
+        scores = F.softmax(cls_score, dim=1).detach()
```
Comment (ZwwWayne) [inline mmdet/models/losses/seesaw_loss.py:58]: how about `scores = F.softmax(cls_score.detach(), dim=1)`

File: mmdet/models/losses/seesaw_loss.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    def _split_cls_score(self, cls_score):
+        # split cls_score to cls_score_classes and cls_score_objectness
+        assert cls_score.size(-1) == self.num_classes + 2
+        cls_score_classes = cls_score[..., :-2]
+        cls_score_objectness = cls_score[..., -2:]
+        return cls_score_classes, cls_score_objectness
+
+    def get_activation(self, cls_score):
+        # calculate the scores for detection results
```
Comment (ZwwWayne) [inline mmdet/models/losses/seesaw_loss.py:147]: We should use standard docstring for `get_activation`, `get_accuracy` and `get_cls_channels` as they will be frequently called in the box heads.

**Review by ZwwWayne — CHANGES_REQUESTED** [review]
We need to add unit tests for seesaw loss and the new functionalities in box head.

**Comment by ZwwWayne** [comment]
PRs can be merged after resolving the comments and fixing unit tests.

File: tests/test_metrics/test_losses.py [hunk 4]
```diff
@@ -109,6 +109,77 @@ def test_kd_loss():
...
+    # only softmax version of Seesaw Loss is implemented
+    with pytest.raises(AssertionError):
+        loss_cfg = dict(type='SeesawLoss', use_sigmoid=True, loss_weight=1.0)
+        build_loss(loss_cfg)
+
+    # test that cls_score.size(-1) == num_classes + 2
+    loss_cls_cfg = dict(
+        type='SeesawLoss', p=0.0, q=0.0, loss_weight=1.0, num_classes=2)
+    loss_cls = build_loss(loss_cls_cfg)
+    with pytest.raises(AssertionError):
```
Comment (ZwwWayne) [inline tests/test_metrics/test_losses.py:122]: Also need to add comments on what error is testing, above this line and line 126.

**Comment by ZwwWayne** [comment]
Need to update https://github.com/open-mmlab/mmdetection/blob/master/docs/model_zoo.md

**Review by ZwwWayne — APPROVED** [review]
